autogen/s3upload.m

```diff
@@ -0,0 +1,51 @@
+% s3upload(mp, bucket, path)
+%
+% This function uploads the artifacts of the last generated model
+%
+% Inputs:
+%   mp, the model parameters struct containing the modelName and modelPath
+%   bucket, the name of the S3 bucket to upload to
+%   path, the path in the S3 bucket that it should be uploaded to
+%   clean, deletes the current files at the path in S3 first before uploading new files
+
+% Copyright 2020 Audio Logic
+%
+% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+% INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+% PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+% FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+% ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+%
+% Dylan Wickham
+% Audio Logic
+% 985 Technology Blvd
+% Bozeman, MT 59718
+% openspeech@flatearthinc.com
+
+function paths = s3upload(mp, bucket, path, clean)
+
+    s3path = 's3://' + string(bucket) + '/' + string(path);
+    disp(s3path)
+    if exist('clean', 'var') && clean
```

Comment:
What's the advantage of this conditional instead of just using `if clean` (assuming `clean` is a bool)?

Reply:
It seems like clean is an optional argument.  Perhaps noting that somewhere would make this a little more clear.  Correct me if I'm wrong, but this is basically saying if clean was passed in and true, then run a command, but if it wasn't passed in, do nothing?

Reply:
That makes sense. I've never seen that pattern for optional arguments before, but I rarely write or read matlab code with optional arguments. I've seen optional arguments handled like this before:
```matlab
if nargin < 4
    clean = false;
end if;
```
So an alternative could be to use `nargin` and assign a default value to clean when it is not included in the function call. That would make it so we don't have to check if `clean` exists. I don't know if either solution is much preferable to the other.

Reply:
I've typically gone with the `nargin` style as well, but I'm not sure if there's some sort of recommended standard for optional arguments.  The `&&` is pretty slick, so I'm good with leaving it as is.  :)

Reply:
I don't have a preference either way. The most popular [MATLAB style guidelines](https://www.mathworks.com/matlabcentral/fileexchange/46056-matlab-style-guidelines-2-0) don't say anything about optional arguments.

Mathworks has a page on [argument validation](https://www.mathworks.com/help/matlab/matlab_prog/function-argument-validation-1.html#mw_43018d69-1d4e-4d7a-813d-b25d64a9db15). They do mention the `nargin` approach, though they spend more time talking about using argument blocks, which are new to me. The argument block can specify the data types of arguments as well as default values for optional positional arguments; they can also handle name-value pair arguments, which are common in official MATLAB functions.

Reply:
I have updated it to use the arguments block syntax which does appear to be pretty clean

Reply:
I do like the arguments block as well. Wish I would have known about it before.